include/cetl/pmr/interface_ptr.hpp

```diff
@@ -0,0 +1,194 @@
+/// @copyright
+/// Copyright (C) OpenCyphal Development Team  <opencyphal.org>
+/// Copyright Amazon.com Inc. or its affiliates.
+/// SPDX-License-Identifier: MIT
+
+#ifndef CETL_PMR_INTERFACE_PTR_H_INCLUDED
+#define CETL_PMR_INTERFACE_PTR_H_INCLUDED
+
+#include "cetl/pmr/function.hpp"
+
+#include <memory>
+#include <functional>
+
+namespace cetl
+{
+namespace pmr
+{
+
+/// RAII helper for cetl::pf17::pmr::polymorphic_allocator and std::pmr::polymorphic_allocator.
+/// Use with cetl::pmr::InterfaceFactory for the best and safest
+/// experience. Remember, be safe, use the cetl::pmr::InterfaceFactory.
+///
+/// @note
+/// See cetl::pmr::InterfaceFactory for an example of how to use this type.
+///
+/// @tparam Interface The interface type of the polymorphic allocator to use for deallocation.
+///
+template <typename Interface>
+class PmrInterfaceDeleter final
+{
+public:
+    /// Constructs a Concrete type-erased deleter for the given interface type.
+    ///
+    /// @tparam PmrAllocator The type of the polymorphic allocator to use for deallocation.
+    ///
+    template <typename PmrAllocator>
+    PmrInterfaceDeleter(PmrAllocator alloc, std::size_t obj_count)
+        : deleter_{[alloc, obj_count](Interface* ptr) mutable {
+            using Concrete = typename PmrAllocator::value_type;
+
+            auto* concrete_ptr = static_cast<Concrete*>(ptr);
+            concrete_ptr->~Concrete();
+            alloc.deallocate(concrete_ptr, obj_count);
+        }}
+    {
+    }
+
+    /// Functor called by smart-pointer to deallocate and deconstruct objects.
+    ///
+    /// @param ptr The object to deconstruct and deallocate.
+    ///
+    void operator()(Interface* ptr) noexcept
+    {
+        deleter_(ptr);
+    }
+
+    // Below convertor constructor is only possible with enabled PMR at `function`.
+    // For now, we decided to comment it out, so that `InterfacePtr` always stays
+    // within 24-bytes small object optimization, namely without extra memory allocation
+    // just for the sake of "advanced" deleter (actually chain of casters down to original `Concrete` pointer).
```

Comment:
```suggestion
    // just for the sake of "advanced" deleter (actually chain of casters down to original `Concrete` pointer).
    //
    // NOTE: It is possible to avoid PMR if we define a certain static limit on the maximum number of
    // down-conversions performed on an interface ptr. One issue is that exceeding that limit is a
    // runtime error, so it will have to be handled correctly somehow. It could be that std::abort()
    // is a sensible choice here, but probably not. The way to implement this solution is to keep a
    // stack of the following converters, where each converter reverses one up-conversion:
    //
    //  using caster_t = void* (*) (void*);
    //  /// Creates a function that performs a safe type conversion on a type-erased void* pointer.
    //  template <typename From, typename To>
    //  static caster_t make_caster() noexcept
    //  { return [](void* const ptr) -> void* { return static_cast<To*>(static_cast<From*>(ptr)); }; }
    //
```